### ISSUE TITLE
Display "name" or "nickname" instead of "first name" + "last name"

### ISSUE THREAD

**Issue by protist** [scheibler/khard#89]
I have some contacts that have the same first and last names and differ only in their middle name. When I attempt to select one from `mutt`'s menu, or the native `khard` menu, the displayed names seem to be derived from the "first name" + "last name" instead of the actual name.

For example, I have the following contact. 

```
Name: John Foo Smith
First name : John
Additional : Foo
Last name  : Smith
```

and also

```
Name: John Bar Smith
First name : John
Additional : Bar
Last name  : Smith
```

From khard's native interface, it's difficult to differentiate between the two.

```
$ khard modify smith
Select contact for Modify action
Address book: owncloud
Index    Name          Phone                 E-Mail                   UID   
1        John Smith    CELL: 1234 567 890    WORK: jbsmith@baz.com    49    
2        John Smith    CELL: 1234 567 891    WORK: jfsmith@baz.com    5a4   
Enter Index:
```

Preferably, the basic contact "name" should be displayed. Another good option would be to display the "nickname" instead, although I guess this would need to be an option since not everyone would like that. Personally, the latter seems useful to me, but the first option is probably "safer".

**Comment by scheibler** [comment]
The "show_nicknames" option already exists in the "contact table" section of the kard config file.
Please have a look into the example config from the misc/khard/ folder and tell me, if that's
helpful. 

I wasn't aware, that contacts with the same first and last name could exist. Maybe I'll add the
middle names to the produced output but then the name column could become relatively large (for
example think about people with three or more middle names).

**Comment by protist** [comment]
> The "show_nicknames" option already exists in the "contact table" section of the kard config file. Please have a look into the example config from the misc/khard/ folder and tell me, if that's helpful.

Ah yes, sorry, I must have missed it. It's definitely helpful for khard itself, although it doesn't affect the selection screen in mutt.

> Maybe I'll add the middle names to the produced output

Perhaps I don't understand the format specification well enough, but is the "main" contact name an independent field, and unrelated to the first/last/middle names? Could khard just not show that field instead?

**Comment by scheibler** [comment]
protist <notifications@github.com> schrieb am 18.08.2016,  1:46 -0700:
>Ah yes, sorry, I must have missed it. It's definitely helpful for khard itself, although it doesn't affect the selection screen in mutt.

Which selection screen do you mean? If you want to add a new email address to an existing contact
(press A) and you produced ambiguous results, the nickname is shown in the upcoming contact
selection table.  If you instead want to write an email and select the receiver address, you don't
see the additional nickname in the name field but the corresponding email address. That should be
enough to distinguish between contacts, cause the email addresses should differ. Am I missing
something?

>> Maybe I'll add the middle names to the produced output
>Perhaps I don't understand the format specification well enough, but is the "main" contact name an independent field, and unrelated to the first/last/middle names? Could khard just not show that field instead?

Khard shows the "fn" (full name, single string) attribute as fallback. Normally the name field of
the contact tables is created programmatically from the "n" attribute (name array).  That's required
for the sort and display by first or last name options (-s and -d). But it wouldn't be a problem to
add the middle names too.

**Comment by protist** [comment]
> If you instead want to write an email and select the receiver address, you don't see the additional nickname in the name field but the corresponding email address. That should be enough to distinguish between contacts, cause the email addresses should differ.

Yes, this is the selection screen I'm talking about. Indeed, the email addresses *do* differ, but in some cases it's still not obvious which contact is which, if the email addresses are, say, `jsmith_22@gmail.com` and `jsmith_1234@gmail.com`.

>Khard shows the "fn" (full name, single string) attribute as fallback.

Do I understand correctly that the "fn" is set by users to be their *preferred* handle for the contact? Hence it makes more sense to display that?

> But it wouldn't be a problem to add the middle names too.

This would be a good alternative if my comment above is incorrect.

**Comment by scheibler** [comment]
>Do I understand correctly that the "fn" is set by users to be their *preferred* handle for the contact? Hence it makes more sense to display that?

The fn (full name) field is just a string which should contain the human readable format of the
contacts name. Problem is, that every contact tool is free to define its own format for that field.
Khard for example uses the format "prefix first additional last suffix" or the organization name, if
no name component is present. Therefore it would not longer be possible to sort your contact table
by last and display by first name as requested in #66.

>> But it wouldn't be a problem to add the middle names too.
>This would be a good alternative if my comment above is incorrect.

Done. Please try out the latest commit from the develop branch.

**Comment by protist** [comment]
>Problem is, that every contact tool is free to define its own format for that field. 

For some reason I had memories of OwnCloud(?) or something else letting you put in arbitrary strings in the "full name" field, and potentially something totally unrelated in the other name fields. You can't modify the "full name" field in the current version, so perhaps my memory is wrong. Are you saying that the "full name" field is programatically derived from the other name fields by the contact tool, or can users manually change it to arbitrary strings?

>Done. Please try out the latest commit from the develop branch.

Works brilliantly in both bare khard and mutt. Thank you for the quick fix.

**Comment by scheibler** [comment]
protist <notifications@github.com> schrieb am 21.08.2016, 19:29 -0700:
>For some reason I had memories of OwnCloud(?) or something else letting you put in arbitrary
>strings in the "full name" field, and potentially something totally unrelated in the other name
>fields. You can't modify the "full name" field in the current version, so perhaps my memory is
>wrong. Are you saying that the "full name" field is programatically derived from the other name
>fields by the contact tool, or can users manually change it to arbitrary strings?

Fn is a standard text attribute, see https://tools.ietf.org/html/rfc6350#section-6.2.1 but it's
mandatory. Khard sets it programmatically to "prefix first middle last suffix" or the organization
name, if the n attribute is absent. So in theory the user could set it on his own but cause it
awaits the full name of the contact, I thought, that would be redundant.

**Comment by protist** [comment]
Ah, right. So khard sets "fn" automatically when you *create* (or modify?) a new contact, but in theory khard might be written to let users define a custom string. In my head, "fn" was a proxy for "preferred name", which is a reason to let users set it manually. Most of my Anglo-Saxon contacts would prefer to be called "First Last", whereas many Chinese contacts prefer "First Middle Last". In addition, if there are two Anglo-Saxon "John Smith"s, then the middle name would again be required.

Having said that, there's certainly an extra layer of complexity involved in allowing users to specify the full name manually, so I'm not sure that it's necessary. I guess it could just be another editable field, though, pre-filled with the "default"? I don't really have a strong feeling either way, to be honest.

**Comment by scheibler** [comment]
protist <notifications@github.com> schrieb am 22.08.2016, 22:39 -0700:
>Ah, right. So khard sets "fn" automatically when you *create* (or modify?) a new contact, but in
>theory khard might be written to let users define a custom string.

Yes.

>Having said that, there's certainly an extra layer of complexity involved in allowing users to
>specify the full name manually, so I'm not sure that it's necessary. I guess it could just be
>another editable field, though, pre-filled with the "default"? I don't really have a strong feeling
>either way, to be honest.

For the output: At the moment khard itself doesn't use the fn attribute at all (only as fallback).
The name column of the contact table either shows "first last" or "last first", depending on the -d
command line option or the display option in the config file. Default is "first last". Now it would
be possible to define a third option for display, let's name it full_name. So you could choose
between first_name, last_name and full_name. But this option would effect the whole contact table.
That means, that you would have to use the full_name option for all contacts and take care of the fn
attribute formatting by yourself (see below). I'am not sure, how many other contact tools really
show the fn attribute. I guess, that most others also do it like khard and derive the name from the
n field.

For the input: The editor could provide a format option for the fn attribute. For example you may
choose from the placeholder keywords PREFIX, FIRST, ADDITIONAL, LAST and SUFFIX and bring them in the order, you
want:

```
full name format : {LAST}, {FIRST} {ADDITIONAL}
```

Default format could come from the config file.

But the implementation is not as trivial as the last one and so its realization depends on how
important that feature is for you. But I think in the long term at least the input part could be
worth the effort.

**Comment by protist** [comment]
I think we are more or less in agreement. Ideally it would be great for users to manually specify fn, but it seems like it's probably too much bother, and adds too much complexity for users. I'm happy with a decision either way. If you'd like to close this issue, then I'm more than happy with that resolution.

**Comment by GSI** [comment]
Regardless of the FN field, it would be nice to just be able to configure a desired format for listings (`khard list`).

Just like in the `date` command where you configure the output freely like `+%Y foo %m bar`, khard could accept formatting options like `%prefix %first %last`.

**Comment by lucc** [comment]
@GSI I still have som old code lying around where I tried to do this (and some more stuff). I will see if I can get it in shape and open a PR (although I am reluctant to open another big PR that will conflict with #92 as I am still waiting for a reaction on that).

**Comment by madduck** [comment]
Hey,

I've also just been "bitten" by this. I have a contact who has a "Prof. Dr." title, but he wants to be called just "Prof." without the "Dr.". Don't ask. ;)

For years, I've thus had

    FN:Prof. John Doe
    N:Doe;John;;Prof. Dr.

in the vcard, and never had a problem with it. `lbdb` would use the `FN` attribute, as would Thunderbird, and others.

The mutt integration of `khard`, however, doesn't use `FN`, but components of `N` instead, with or without the nickname. I am not sure this is desirable.

Instead, I think it's correct that `khard` auto-assembles `FN` from the components of `N` on creation, but then I think it should favour `FN` over `N`, which would allow a user to modify the desired representation of a name, without changing the components data.

Should I create another issue about this and let this issue be about how to programmatically fill `FN`, which is still useful on creation and all the better if configurable?

**Comment by lucc** [comment]
We already have a `show_nicknames` and the `--display` and `--sort` options.

In #222 I propose to make `FN` editable. If it is empty it will be autofilled like before. This would mean that `FN` can *not* be configured with variables or a patter but can be editied for each individual contact.
@protist @madduck  does that solve your issues?

**Comment by protist** [comment]
That sounds fantastic. Thank you. I'm happy for this to be closed.

**Comment by lucc** [comment]
closed by #222
